File: app/src/main/java/be/digitalia/fosdem/fragments/EventDetailsFragment.java
package be.digitalia.fosdem.fragments;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.provider.CalendarContract;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.TextPaint;
import android.text.TextUtils;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.google.android.material.snackbar.Snackbar;

import java.text.DateFormat;
import java.util.Date;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.core.app.ShareCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import be.digitalia.fosdem.R;
import be.digitalia.fosdem.activities.PersonInfoActivity;
import be.digitalia.fosdem.api.FosdemApi;
import be.digitalia.fosdem.model.Building;
import be.digitalia.fosdem.model.Event;
import be.digitalia.fosdem.model.EventDetails;
import be.digitalia.fosdem.model.Link;
import be.digitalia.fosdem.model.Person;
import be.digitalia.fosdem.model.RoomStatus;
import be.digitalia.fosdem.utils.ClickableArrowKeyMovementMethod;
import be.digitalia.fosdem.utils.CustomTabsUtils;
import be.digitalia.fosdem.utils.DateUtils;
import be.digitalia.fosdem.utils.StringUtils;
import be.digitalia.fosdem.viewmodels.EventDetailsViewModel;

public class EventDetailsFragment extends Fragment {

	static class ViewHolder {
		LayoutInflater inflater;
		TextView personsTextView;
		TextView roomStatus;
		View linksHeader;
		ViewGroup linksContainer;
	}

	private static final String ARG_EVENT = "event";

	Event event;
	ViewHolder holder;
	EventDetailsViewModel viewModel;

	public static EventDetailsFragment newInstance(Event event) {
		EventDetailsFragment f = new EventDetailsFragment();
		Bundle args = new Bundle();
		args.putParcelable(ARG_EVENT, event);
		f.setArguments(args);
		return f;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		event = requireArguments().getParcelable(ARG_EVENT);
		viewModel = new ViewModelProvider(this).get(EventDetailsViewModel.class);
		viewModel.setEvent(event);
		setHasOptionsMenu(true);
	}

	public Event getEvent() {
		return event;
	}

	@Override
	public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_event_details, container, false);

		holder = new ViewHolder();
		holder.inflater = inflater;

		((TextView) view.findViewById(R.id.title)).setText(event.getTitle());
		TextView textView = view.findViewById(R.id.subtitle);
		String text = event.getSubTitle();
		if (TextUtils.isEmpty(text)) {
			textView.setVisibility(View.GONE);
		} else {
			textView.setText(text);
		}

		// Set the persons summary text first; replace it with the clickable text when the loader completes
		holder.personsTextView = view.findViewById(R.id.persons);
		String personsSummary = event.getPersonsSummary();
		if (TextUtils.isEmpty(personsSummary)) {
			holder.personsTextView.setVisibility(View.GONE);
		} else {
			holder.personsTextView.setText(personsSummary);
			holder.personsTextView.setMovementMethod(LinkMovementMethod.getInstance());
			holder.personsTextView.setVisibility(View.VISIBLE);
		}


		textView = view.findViewById(R.id.time);
		Date startTime = event.getStartTime();
		Date endTime = event.getEndTime();
		DateFormat timeDateFormat = DateUtils.getTimeDateFormat(getActivity());
		text = String.format("%1$s, %2$s ― %3$s",
				event.getDay().toString(),
				(startTime != null) ? timeDateFormat.format(startTime) : "?",
				(endTime != null) ? timeDateFormat.format(endTime) : "?");
		textView.setText(text);
		textView.setContentDescription(getString(R.string.time_content_description, text));

		textView = view.findViewById(R.id.room);
		final String roomName = event.getRoomName();
		Spannable roomText = new SpannableString(String.format("%1$s (Building %2$s)", roomName, Building.fromRoomName(roomName)));
		final int roomImageResId = getResources().getIdentifier(StringUtils.roomNameToResourceName(roomName), "drawable", requireActivity().getPackageName());
		// If the room image exists, make the room text clickable to display it
		if (roomImageResId != 0) {
			roomText.setSpan(new ClickableSpan() {
				@Override
				public void onClick(@NonNull View view) {
					RoomImageDialogFragment.newInstance(roomName, roomImageResId).show(getParentFragmentManager());
				}

				@Override
				public void updateDrawState(@NonNull TextPaint ds) {
					super.updateDrawState(ds);
					ds.setUnderlineText(false);
				}
			}, 0, roomText.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
			textView.setMovementMethod(LinkMovementMethod.getInstance());
		}
		textView.setText(roomText);
		textView.setContentDescription(getString(R.string.room_content_description, roomText));

		holder.roomStatus = view.findViewById(R.id.room_status);


		textView = view.findViewById(R.id.abstract_text);
		text = event.getAbstractText();
		if (TextUtils.isEmpty(text)) {
			textView.setVisibility(View.GONE);
		} else {
			textView.setText(StringUtils.parseHtml(text, getResources()));
			textView.setMovementMethod(ClickableArrowKeyMovementMethod.getInstance());
		}
		textView = view.findViewById(R.id.description);
		text = event.getDescription();
		if (TextUtils.isEmpty(text)) {
			textView.setVisibility(View.GONE);
		} else {
			textView.setText(StringUtils.parseHtml(text, getResources()));
			textView.setMovementMethod(ClickableArrowKeyMovementMethod.getInstance());
		}

		holder.linksHeader = view.findViewById(R.id.links_header);
		holder.linksContainer = view.findViewById(R.id.links_container);
		return view;
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		viewModel.getEventDetails().observe(getViewLifecycleOwner(), eventDetails -> {
			if (eventDetails != null) {
				setEventDetails(eventDetails);
			}
		});

		// Live room status
		FosdemApi.getRoomStatuses(requireContext()).observe(getViewLifecycleOwner(), roomStatuses -> {
			RoomStatus roomStatus = roomStatuses.get(event.getRoomName());
			if (roomStatus == null) {
				holder.roomStatus.setText(null);
			} else {
				holder.roomStatus.setText(roomStatus.getNameResId());
				holder.roomStatus.setTextColor(ContextCompat.getColorStateList(requireContext(), roomStatus.getColorResId()));
			}
		});
	}

	@Override
	public void onDestroyView() {
		super.onDestroyView();
		holder = null;
	}

	@Override
	public void onCreateOptionsMenu(@NonNull Menu menu, MenuInflater inflater) {
		inflater.inflate(R.menu.event, menu);
		menu.findItem(R.id.share).setIntent(getShareChooserIntent());
	}

	private Intent getShareChooserIntent() {
		return ShareCompat.IntentBuilder.from(requireActivity())
				.setSubject(String.format("%1$s (FOSDEM)", event.getTitle()))
				.setType("text/plain")
				.setText(String.format("%1$s %2$s #FOSDEM", event.getTitle(), event.getUrl()))
				.setChooserTitle(R.string.share)
				.createChooserIntent();
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.add_to_agenda:
				addToAgenda();
				return true;
		}
		return false;
	}

	private void addToAgenda() {
		Intent intent = new Intent(Intent.ACTION_EDIT);
		intent.setType("vnd.android.cursor.item/event");
		intent.putExtra(CalendarContract.Events.TITLE, event.getTitle());
		intent.putExtra(CalendarContract.Events.EVENT_LOCATION, "ULB - " + event.getRoomName());
		String description = event.getAbstractText();
		if (TextUtils.isEmpty(description)) {
			description = event.getDescription();
		}
		description = StringUtils.stripHtml(description);
		// Add speaker info if available
		EventDetails details = viewModel.getEventDetails().getValue();
		final int personsCount = (details == null) ? 0 : details.getPersons().size();
		if (personsCount > 0) {
			description = String.format("%1$s: %2$s\n\n%3$s", getResources().getQuantityString(R.plurals.speakers, personsCount), event.getPersonsSummary(), description);
		}
		intent.putExtra(CalendarContract.Events.DESCRIPTION, description);
		Date time = event.getStartTime();
		if (time != null) {
			intent.putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, time.getTime());
		}
		time = event.getEndTime();
		if (time != null) {
			intent.putExtra(CalendarContract.EXTRA_EVENT_END_TIME, time.getTime());
		}
		try {
			startActivity(intent);
		} catch (ActivityNotFoundException e) {
			Snackbar.make(requireView(), R.string.calendar_not_found, Snackbar.LENGTH_LONG).show();
		}
	}

	void setEventDetails(@NonNull EventDetails eventDetails) {
		// 1. Persons
		final List<Person> persons = eventDetails.getPersons();
		if (persons.size() > 0) {
			// Build a list of clickable persons
			SpannableStringBuilder sb = new SpannableStringBuilder();
			int length = 0;
			for (Person person : persons) {
				if (length != 0) {
					sb.append(", ");
				}
				String name = person.getName();
				sb.append(name);
				length = sb.length();
				sb.setSpan(new PersonClickableSpan(person), length - name.length(), length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
			}
			holder.personsTextView.setText(sb);
			holder.personsTextView.setVisibility(View.VISIBLE);
		}

		// 2. Links
		final List<Link> links = eventDetails.getLinks();
		holder.linksContainer.removeAllViews();
		if (links.size() > 0) {
			holder.linksHeader.setVisibility(View.VISIBLE);
			holder.linksContainer.setVisibility(View.VISIBLE);
			for (Link link : links) {
				View view = holder.inflater.inflate(R.layout.item_link, holder.linksContainer, false);
				TextView tv = view.findViewById(R.id.description);
				tv.setText(link.getDescription());
				view.setOnClickListener(new LinkClickListener(event, link));
				holder.linksContainer.addView(view);
			}
		} else {
			holder.linksHeader.setVisibility(View.GONE);
			holder.linksContainer.setVisibility(View.GONE);
		}
	}

	private static class PersonClickableSpan extends ClickableSpan {

		private final Person person;

		PersonClickableSpan(Person person) {
			this.person = person;
		}

		@Override
		public void onClick(@NonNull View v) {
			Context context = v.getContext();
			Intent intent = new Intent(context, PersonInfoActivity.class).putExtra(PersonInfoActivity.EXTRA_PERSON, person);
			context.startActivity(intent);
		}

		@Override
		public void updateDrawState(@NonNull TextPaint ds) {
			super.updateDrawState(ds);
			ds.setUnderlineText(false);
		}
	}

	private static class LinkClickListener implements View.OnClickListener {

		private final Event event;
		private final Link link;

		LinkClickListener(@NonNull Event event, @NonNull Link link) {
			this.event = event;
			this.link = link;
		}

		@Override
		public void onClick(View v) {
			String url = link.getUrl();
			try {
				final Context context = v.getContext();
				CustomTabsUtils.configureToolbarColors(new CustomTabsIntent.Builder(), context, event.getTrack().getType().getAppBarColorResId())
						.setShowTitle(true)
						.setStartAnimations(context, R.anim.slide_in_right, R.anim.slide_out_left)
						.setExitAnimations(context, R.anim.slide_in_left, R.anim.slide_out_right)
						.build()
						.launchUrl(context, Uri.parse(url));
			} catch (ActivityNotFoundException ignore) {
			}
		}
	}
}

File: app/src/main/java/be/digitalia/fosdem/fragments/EventDetailsFragment.kt
package be.digitalia.fosdem.fragments

import android.annotation.SuppressLint
import android.content.ActivityNotFoundException
import android.content.Intent
import android.os.Bundle
import android.provider.CalendarContract
import android.text.SpannableString
import android.text.TextPaint
import android.text.style.ClickableSpan
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.annotation.LayoutRes
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.app.ShareCompat
import androidx.core.content.ContextCompat
import androidx.core.net.toUri
import androidx.core.text.buildSpannedString
import androidx.core.text.inSpans
import androidx.core.text.method.LinkMovementMethodCompat
import androidx.core.text.set
import androidx.core.view.MenuProvider
import androidx.core.view.isVisible
import androidx.core.view.plusAssign
import androidx.fragment.app.Fragment
import androidx.fragment.app.add
import androidx.fragment.app.commit
import androidx.fragment.app.viewModels
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.withStarted
import be.digitalia.fosdem.R
import be.digitalia.fosdem.activities.PersonInfoActivity
import be.digitalia.fosdem.api.FosdemApi
import be.digitalia.fosdem.model.Building
import be.digitalia.fosdem.model.Event
import be.digitalia.fosdem.model.EventDetails
import be.digitalia.fosdem.model.Person
import be.digitalia.fosdem.model.Resource
import be.digitalia.fosdem.model.RoomStatus
import be.digitalia.fosdem.settings.UserSettingsProvider
import be.digitalia.fosdem.utils.ClickableArrowKeyMovementMethod
import be.digitalia.fosdem.utils.DateUtils
import be.digitalia.fosdem.utils.configureToolbarColors
import be.digitalia.fosdem.utils.getParcelableCompat
import be.digitalia.fosdem.utils.launchAndRepeatOnLifecycle
import be.digitalia.fosdem.utils.parseHtml
import be.digitalia.fosdem.utils.roomNameToResourceName
import be.digitalia.fosdem.utils.stripHtml
import be.digitalia.fosdem.utils.toLocalDateTimeOrNull
import be.digitalia.fosdem.viewmodels.EventDetailsViewModel
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.lifecycle.withCreationCallback
import kotlinx.coroutines.launch
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import javax.inject.Inject

@AndroidEntryPoint
class EventDetailsFragment : Fragment(R.layout.fragment_event_details) {

    private class ViewHolder(view: View) {
        val personsTextView: TextView = view.findViewById(R.id.persons)
        val timeTextView: TextView = view.findViewById(R.id.time)
        val roomStatusTextView: TextView = view.findViewById(R.id.room_status)
        val attachmentsHeader: View = view.findViewById(R.id.attachments_header)
        val attachmentsContainer: ViewGroup = view.findViewById(R.id.attachments_container)
        val linksHeader: View = view.findViewById(R.id.links_header)
        val linksContainer: ViewGroup = view.findViewById(R.id.links_container)
        val resourcesFooter: View = view.findViewById(R.id.resources_footer)
    }

    @Inject
    lateinit var userSettingsProvider: UserSettingsProvider
    @Inject
    lateinit var api: FosdemApi
    private val viewModel: EventDetailsViewModel by viewModels(extrasProducer = {
        defaultViewModelCreationExtras.withCreationCallback<EventDetailsViewModel.Factory> { factory ->
            factory.create(event)
        }
    })

    val event by lazy<Event>(LazyThreadSafetyMode.NONE) {
        requireArguments().getParcelableCompat(ARG_EVENT)!!
    }

    private inner class EventDetailsMenuProvider : MenuProvider {
        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            menuInflater.inflate(R.menu.event, menu)
            menu.findItem(R.id.share)?.intent = createShareChooserIntent()
        }

        private fun createShareChooserIntent(): Intent {
            val title = event.title.orEmpty()
            val url = event.url.orEmpty()
            return ShareCompat.IntentBuilder(requireContext())
                .setSubject("$title ($CONFERENCE_NAME)")
                .setType("text/plain")
                .setText("$title $url $CONFERENCE_HASHTAG")
                .setChooserTitle(R.string.share)
                .createChooserIntent()
        }

        override fun onMenuItemSelected(menuItem: MenuItem) = when (menuItem.itemId) {
            R.id.add_to_agenda -> {
                addToAgenda()
                true
            }
            else -> false
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        requireActivity().addMenuProvider(EventDetailsMenuProvider(), viewLifecycleOwner)

        val timeFormatter = DateUtils.getTimeFormatter(view.context)

        val holder = ViewHolder(view).apply {
            view.findViewById<TextView>(R.id.title).text = event.title
            view.findViewById<TextView>(R.id.subtitle).apply {
                val subTitle = event.subTitle
                if (subTitle.isNullOrEmpty()) {
                    isVisible = false
                } else {
                    text = subTitle
                }
            }

            personsTextView.apply {
                // Set the persons summary text first;
                // replace it with the clickable text when the event details loading completes
                val personsSummary = event.personsSummary
                if (personsSummary.isNullOrEmpty()) {
                    isVisible = false
                } else {
                    text = personsSummary
                    movementMethod = LinkMovementMethodCompat.getInstance()
                    isVisible = true
                }
            }

            // Display time placeholder until the ZoneId is loaded from user preferences
            bindTime(timeTextView, timeFormatter, null)

            view.findViewById<TextView>(R.id.room).apply {
                val roomName = event.roomName
                if (roomName.isNullOrEmpty()) {
                    isVisible = false
                } else {
                    val building = Building.fromRoomName(roomName)
                    val roomText = SpannableString(if (building == null) roomName else getString(R.string.room_building, roomName, building))
                    val roomImageResId = resources.getIdentifier(roomNameToResourceName(roomName), "drawable", requireActivity().packageName)

                    // Make the room text clickable to display the room image, even if it doesn't exist
                    roomText[0, roomText.length] = object : ClickableSpan() {
                        override fun onClick(view: View) {
                            parentFragmentManager.commit(allowStateLoss = true) {
                                add<RoomImageDialogFragment>(RoomImageDialogFragment.TAG,
                                        args = RoomImageDialogFragment.createArguments(roomName, roomImageResId))
                            }
                        }

                        override fun updateDrawState(ds: TextPaint) {
                            super.updateDrawState(ds)
                            ds.isUnderlineText = false
                        }
                    }
                    movementMethod = LinkMovementMethodCompat.getInstance()

                    text = roomText
                    contentDescription = getString(R.string.room_content_description, roomText)
                }
            }

            view.findViewById<TextView>(R.id.abstract_text).apply {
                val abstractText = event.abstractText
                if (abstractText.isNullOrEmpty()) {
                    isVisible = false
                } else {
                    text = abstractText.parseHtml(resources)
                    movementMethod = ClickableArrowKeyMovementMethod
                }
            }

            view.findViewById<TextView>(R.id.description).apply {
                val descriptionText = event.description
                if (descriptionText.isNullOrEmpty()) {
                    isVisible = false
                } else {
                    text = descriptionText.parseHtml(resources)
                    movementMethod = ClickableArrowKeyMovementMethod
                }
            }
        }

        viewLifecycleOwner.lifecycleScope.launch {
            val eventDetails = viewModel.eventDetails.await()
            viewLifecycleOwner.withStarted {
                showEventDetails(holder, eventDetails)
            }
        }

        val roomName = event.roomName
        viewLifecycleOwner.launchAndRepeatOnLifecycle {
            launch {
                userSettingsProvider.zoneId.collect { zoneId ->
                    bindTime(holder.timeTextView, timeFormatter, zoneId)
                }
            }

            // Live room status
            if (!roomName.isNullOrEmpty()) {
                launch {
                    api.roomStatuses.collect { statuses ->
                        bindRoomStatus(holder.roomStatusTextView, statuses[roomName])
                    }
                }
            }
        }
    }

    @SuppressLint("SetTextI18n")
    private fun bindTime(timeTextView: TextView, timeFormatter: DateTimeFormatter, zoneId: ZoneId?) {
        val startTime = event.startTime?.toLocalDateTimeOrNull(zoneId)?.format(timeFormatter) ?: "?"
        val endTime = event.endTime?.toLocalDateTimeOrNull(zoneId)?.format(timeFormatter) ?: "?"
        timeTextView.text = "${event.day}, $startTime ― $endTime"
        timeTextView.contentDescription = getString(R.string.time_content_description, timeTextView.text)
    }

    private fun bindRoomStatus(roomStatusTextView: TextView, roomStatus: RoomStatus?) {
        if (roomStatus == null) {
            roomStatusTextView.text = null
        } else {
            roomStatusTextView.setText(roomStatus.nameResId)
            roomStatusTextView.setTextColor(
                ContextCompat.getColorStateList(roomStatusTextView.context, roomStatus.colorResId)
            )
        }
    }

    private fun addToAgenda() {
        val intent = Intent(Intent.ACTION_EDIT).apply {
            type = "vnd.android.cursor.item/event"
            event.title?.let { putExtra(CalendarContract.Events.TITLE, it) }
            val roomName = event.roomName
            val location = if (roomName.isNullOrEmpty()) VENUE_NAME else "$VENUE_NAME - $roomName"
            putExtra(CalendarContract.Events.EVENT_LOCATION, location)

            var description = event.abstractText
            if (description.isNullOrEmpty()) {
                description = event.description.orEmpty()
            }
            description = description.stripHtml()
            // Add speaker info if available
            val personsSummary = event.personsSummary
            if (!personsSummary.isNullOrBlank()) {
                val personsCount = personsSummary.count { it == ',' } + 1
                val speakersLabel = resources.getQuantityString(R.plurals.speakers, personsCount)
                description = "$speakersLabel: $personsSummary\n\n$description"
            }
            putExtra(CalendarContract.Events.DESCRIPTION, description)
            event.startTime?.let { putExtra(CalendarContract.EXTRA_EVENT_BEGIN_TIME, it.toEpochMilli()) }
            event.endTime?.let { putExtra(CalendarContract.EXTRA_EVENT_END_TIME, it.toEpochMilli()) }
        }

        try {
            startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            Snackbar.make(requireView(), R.string.calendar_not_found, Snackbar.LENGTH_LONG).show()
        }
    }

    private fun showEventDetails(holder: ViewHolder, eventDetails: EventDetails) {
        holder.run {
            val (persons, attachments, links) = eventDetails

            // 1. Persons
            if (persons.isNotEmpty()) {
                // Build a list of clickable persons
                val clickablePersonsSummary = buildSpannedString {
                    for (person in persons) {
                        val name = person.name
                        if (name.isNullOrEmpty()) {
                            continue
                        }
                        if (isNotEmpty()) {
                            append(", ")
                        }
                        inSpans(PersonClickableSpan(person)) {
                            append(name)
                        }
                    }
                }
                personsTextView.text = clickablePersonsSummary
                personsTextView.isVisible = true
            }

            // 2. Attachments
            populateResources(attachments, R.layout.item_attachment, attachmentsHeader, attachmentsContainer)

            // 3. Links
            populateResources(links, R.layout.item_link, linksHeader, linksContainer)

            resourcesFooter.isVisible = attachments.isNotEmpty() || links.isNotEmpty()
        }
    }

    private fun populateResources(
        resources: List<Resource>,
        @LayoutRes layoutResId: Int,
        header: View,
        container: ViewGroup
    ) {
        container.removeAllViews()
        if (resources.isNotEmpty()) {
            header.isVisible = true
            container.isVisible = true
            val inflater = layoutInflater
            for (resource in resources) {
                val view = inflater.inflate(layoutResId, container, false)
                view.findViewById<TextView>(R.id.description).apply {
                    text = resource.description
                }
                view.setOnClickListener(ResourceClickListener(event, resource))
                container += view
            }
        } else {
            header.isVisible = false
            container.isVisible = false
        }
    }

    private class PersonClickableSpan(private val person: Person) : ClickableSpan() {
        override fun onClick(v: View) {
            val context = v.context
            val intent = Intent(context, PersonInfoActivity::class.java)
                    .putExtra(PersonInfoActivity.EXTRA_PERSON, person)
            context.startActivity(intent)
        }

        override fun updateDrawState(ds: TextPaint) {
            super.updateDrawState(ds)
            ds.isUnderlineText = false
        }
    }

    private class ResourceClickListener(
        private val event: Event,
        private val resource: Resource
    ) : View.OnClickListener {
        override fun onClick(v: View) {
            try {
                val context = v.context
                CustomTabsIntent.Builder()
                        .configureToolbarColors(context, event.track.type.appBarColorResId)
                        .setShowTitle(true)
                        .setStartAnimations(context, R.anim.slide_in_right, R.anim.slide_out_left)
                        .setExitAnimations(context, R.anim.slide_in_left, R.anim.slide_out_right)
                        .build()
                        .launchUrl(context, resource.url.toUri())
            } catch (ignore: ActivityNotFoundException) {
            }
        }
    }

    companion object {
        private const val ARG_EVENT = "event"
        private const val CONFERENCE_NAME = "FOSDEM"
        private const val CONFERENCE_HASHTAG = "#FOSDEM"
        private const val VENUE_NAME = "ULB"

        fun createArguments(event: Event) = Bundle(1).apply {
            putParcelable(ARG_EVENT, event)
        }
    }
}